src/Stratis.Bitcoin.Features.BlockStore/BlockStoreSignaled.cs

```diff
@@ -104,17 +118,56 @@ protected override void OnNextCore(Block block)
             this.logger.LogTrace("(-)");
         }
 
+        /// <summary>Initializes the <see cref="BlockStoreSignaled"/>.</summary>
+        public void Initialize()
+        {
+            this.batchTimer.Elapsed += async (sender, args) => { await this.OnBatchTimerRunsOutAsync().ConfigureAwait(false); };
```

Comment:
Ah one more question I am not sure about will the timer elapse even that the previous call did not complete?
Say timer is 1 sec and a call is taking 1.1 sec will the next call queue a task or will the timer wait time when the previous call completed.

Reply:
```
            // Init timer with 5ms interval that call task that does the Delay with 1 sec.
            System.Timers.Timer timer = new System.Timers.Timer(5) { AutoReset = false }; ;
            timer.Elapsed += async (sender, eventArgs) =>
            {
                await Task.Delay(1000);

                Console.WriteLine("CALL");
            };

            // Start timer 4 times with 100ms interval.
            for (int i = 0; i < 4; ++i)
            {
                timer.Start();
                Thread.Sleep(100);
            }

            Thread.Sleep(10000);
            Console.ReadLine();
```

This produces this console output:

```
CALL
CALL
CALL
CALL
```

So it looks like timer will still invoke the callback even if the previous one is not completed.

Reply:
Yes that's the difference with using a timer we don't wait completion.
Tasks will be waiting at the async lock indeed, I guess I am fine with that.
Would be nice to have a Timer that wait completion (perhaps we can modify the AsyncLoop to have an on/off button but it might be tricky)

Reply:
i haven't checked the code yet, but the discussion about the timer here seems unrelated to what we do (or should do) in this code, well i'll check it after lunch and see myself

Reply:
>Yes that's the difference with using a timer we don't wait completion.

In this particular case it's not that bad and I can't imagine a scenario when we will have 3 callbacks running at the same time. Actually there is just 1 occasion when we can get 2 of them queued: when we're are not synced and it's the 1st time we get a tip block. So just once per node run we can end up in scenario when we can have 2 callbacks invoked at the same time.

Reply:
Fair enough as we will be disabling the timer once we reached the tip.